### PR TITLE
feat: :sparkles: implement upsert user preference, update logic for user login

### PR DESCRIPTION
**Pull request by suryanirvana**
### Background: 
Enables user to add their preferences when login for the first time

### Additional Packages/Dependencies:
none

### Changes:
- /auth/entity/user.entity.ts
- /auth/service/auth.service.ts

- /common/middleware/authorization.middleware.ts (NEW)
- /common/service/base.service.ts
- /common/utils/jwt.ts

- /user/controller/user.controller.ts (NEW)
- /user/dto/user.preference.dto.ts (NEW)
- /user/entity/preference.entity.ts (NEW)
- /user/entity/preference.entity.ts (NEW)
- /user/request/upsertUserPreference.request.ts (NEW)
- /user/response/upsertUserPreference.response.ts (NEW)
- /user/service/user.service.ts (NEW)

- app.ts

### How to Implement:
- [x] Create necessary files
- [x] Implement authorization middleware, where it throws error when making a request with no auth token or invalid token
- [x] Update logic for user login, to update isFirstLogin
- [x] Implement upsert user preferences by getting the old preferences (if exist, to avoid error), removing it, and then adding the new preferences

### CODE REVIEW

File: src/user/service/user.service.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        try {
+            const preferenceID = body.preferenceID;
+            const preferences = await this.getPreferences(preferenceID);
+            if (!preferences) throw new BadRequestException('Preferences Invalid');
+
+            const email = locals.email;
+            const username = locals.username;
+            const user = await this.getUserByEmailAndUsername(email, username);
+            if (!user) throw new NotFoundException('User does not exist.');
+            if (!user.isVerified)
```
Comment (RafadanaM) [inline src/user/service/user.service.ts]: This check can be moved to auth middleware and probably store whether the user is verified or not on jwt

**Review by RafadanaM — CHANGES_REQUESTED** [review]
Please check the comments

File: src/auth/service/auth.service.ts [hunk 2]
```diff
@@ -55,21 +55,7 @@ class AuthService implements BaseService {
 
     public login = async (body: LoginDTO) => {
```
Comment (RafadanaM) [inline src/auth/service/auth.service.ts:57]: please send user preferences too

File: src/user/service/user.service.ts (hunk 1 above)
Comment (suryanirvana) [reply src/user/service/user.service.ts]: Do kindly check the latest commit 🙂

File: src/auth/service/auth.service.ts (hunk 2 above)
Comment (suryanirvana) [reply src/auth/service/auth.service.ts:57]: Do kindly check the latest commit 🙂

**Review by RafadanaM — APPROVED** [review]

**Pull request merged**